plugins/notifications-backend/src/service/DefaultNotificationRecipientResolver.ts

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Entity,
+  isGroupEntity,
+  isUserEntity,
+  parseEntityRef,
+  RELATION_HAS_MEMBER,
+  RELATION_OWNED_BY,
+  RELATION_PARENT_OF,
+  stringifyEntityRef,
+} from '@backstage/catalog-model';
+import { AuthService } from '@backstage/backend-plugin-api';
+import { CatalogService } from '@backstage/plugin-catalog-node';
+import { NotificationRecipientResolver } from '@backstage/plugin-notifications-node';
+
+const isUserEntityRef = (ref: string) =>
+  parseEntityRef(ref).kind.toLocaleLowerCase() === 'user';
+
+// Partitions array of entity references to two arrays; user entity refs and other entity refs
+const partitionEntityRefs = (refs: string[]): string[][] => {
+  const ret = [[], []] as string[][];
+  for (const ref of refs) {
+    if (isUserEntityRef(ref)) {
+      ret[0].push(ref);
+    } else {
+      ret[1].push(ref);
+    }
+  }
+  return ret;
+};
+
+export class DefaultNotificationRecipientResolver
+  implements NotificationRecipientResolver
+{
+  constructor(
+    private readonly auth: AuthService,
+    private readonly catalog: CatalogService,
+  ) {}
+
+  async resolveNotificationRecipients(options: {
+    entityRefs: string[];
+    excludeEntityRefs?: string[];
+  }): Promise<{ userEntityRefs: string[] }> {
+    const { entityRefs, excludeEntityRefs = [] } = options;
+
+    const [userEntityRefs, otherEntityRefs] = partitionEntityRefs(entityRefs);
+    const users: string[] = userEntityRefs.filter(
+      ref => !excludeEntityRefs.includes(ref),
+    );
+    const filtered = otherEntityRefs.filter(
+      ref => !excludeEntityRefs.includes(ref),
+    );
+
+    const fields = ['kind', 'metadata.name', 'metadata.namespace', 'relations'];
+    let entities: Array<Entity | undefined> = [];
+    if (filtered.length > 0) {
+      const fetchedEntities = await this.catalog.getEntitiesByRefs(
+        {
+          entityRefs: filtered,
+          fields,
+        },
+        { credentials: await this.auth.getOwnServiceCredentials() },
+      );
+      entities = fetchedEntities.items;
+    }
+
+    const cachedEntityRefs = new Map<string, string[]>();
+
+    const mapEntity = async (entity: Entity | undefined): Promise<string[]> => {
+      if (!entity) {
+        return [];
+      }
+
+      const currentEntityRef = stringifyEntityRef(entity);
```

Comment:
I'll just note that this array is the same length and in the same order as the input (`filtered `), so the refs are already available while iterating (if grabbing the indices as you go)). But no worries